test/util/dynamic/dynamic.go

```diff
@@ -0,0 +1,89 @@
+package dynamic
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// Client returns a client from a given unstructured object.
+// It can be used when creating an object from a yaml file.
+type Client interface {
+	GetClient(obj *unstructured.Unstructured) (ResourceClient, error)
+}
+
+type client struct {
+	dynamic.Interface
+	mapping meta.RESTMapper
+}
+
+func NewDynamicClient(kubeConfig *rest.Config) (Client, error) {
+	cli, err := dynamic.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	groupResources, err := restmapper.GetAPIGroupResources(discoveryClient)
+	if err != nil {
+		return nil, err
+	}
+	mapper := restmapper.NewDiscoveryRESTMapper(groupResources)
+
+	return &client{cli, mapper}, nil
+}
+
+func (d *client) getGVR(obj *unstructured.Unstructured) (*meta.RESTMapping, error) {
+	var gvk = obj.GroupVersionKind()
+	return d.mapping.RESTMapping(gvk.GroupKind(), gvk.Version)
+}
+
+// GetClient returns a non-namespaced or namespaced ResourceClient depending on a given object.
+func (d *client) GetClient(obj *unstructured.Unstructured) (ResourceClient, error) {
+	gvr, err := d.getGVR(obj)
+	if err != nil {
+		return nil, err
+	}
+	if obj.GetNamespace() != "" {
+		return &resourceClient{d.Resource(gvr.Resource).Namespace(obj.GetNamespace())}, nil
+	}
+	return &resourceClient{d.Resource(gvr.Resource)}, nil
+}
+
+// ResourceClient is an interface that can be used for *K8sObjectWithRetry helper functions.
+// The default dynamic ResourceInterface can't be used for *K8sObjectWithRetry
+// because it has different method signatures from other clients interface.
```

Comment:
nit: it would be clearer to mention that the reason for this diff is thata dynamic clients support actions over any subresource, whereas the typed clients break those out into extensions.